src/mainboard/starfive/visionfive2/bt0/src/init.rs

```diff
@@ -0,0 +1,548 @@
+pub const SYSCON_SYSMAIN_CTRL28: u32 = SYSCON_SYSMAIN_CTRL_BASE + 0x0070;
+pub const SYSCON_SYSMAIN_CTRL68: u32 = SYSCON_SYSMAIN_CTRL_BASE + 0x0110;
+pub const SYSCON_SYSMAIN_CTRL69: u32 = SYSCON_SYSMAIN_CTRL_BASE + 0x0114;
+
+// SYSCON_SYSMAIN_PLLx [31:24] [23:16] [15:8]   [7:4] [3]    [2]   [1] [0]
+//                      OD      BWADJ  CLKFDIV  CLKR  bypass infb  pd  rst
+
+pub fn syscon_pll1_reset(r: bool) {
+    if r {
+        write_32(SYSCON_SYSMAIN_PLL1, 0x0029_2905);
+    } else {
+        write_32(SYSCON_SYSMAIN_PLL1, 0x0029_2904);
+    }
+}
+
+pub fn syscon_gmac_phy_intf_sel(v: u32) {
+    let nv = read_32(SYSCON_SYSMAIN_CTRL28) & !(0x7);
+    write_32(SYSCON_SYSMAIN_CTRL28, nv | (v & 0x7));
+}
+
+pub fn disable_u74_memaxi_remap(v: u32) {
+    let nv = read_32(SYSCON_SYSMAIN_CTRL68) & !(0x1);
+    write_32(SYSCON_SYSMAIN_CTRL68, nv | (v & 0x1));
+}
+
+pub fn syscon_core1_en(v: u32) {
+    let nv = read_32(SYSCON_SYSMAIN_CTRL69) & !(0x1);
+    write_32(SYSCON_SYSMAIN_CTRL69, nv | (v & 0x1));
+}
+
+pub const SYSCON_IOPAD_CTRL_BASE: u32 = 0x00_1185_8000;
+pub const SYSCON_IOPAD_CTRL32: u32 = SYSCON_IOPAD_CTRL_BASE + 0x80;
+pub const SYSCON_IOPAD_CTRL33: u32 = SYSCON_IOPAD_CTRL_BASE + 0x84;
+pub const SYSCON_IOPAD_CTRL34: u32 = SYSCON_IOPAD_CTRL_BASE + 0x88;
+pub const SYSCON_IOPAD_CTRL35: u32 = SYSCON_IOPAD_CTRL_BASE + 0x8c;
+pub const SYSCON_IOPAD_CTRL38: u32 = SYSCON_IOPAD_CTRL_BASE + 0x98;
+pub const SYSCON_IOPAD_CTRL39: u32 = SYSCON_IOPAD_CTRL_BASE + 0x9C;
+pub const SYSCON_IOPAD_CTRL50: u32 = SYSCON_IOPAD_CTRL_BASE + 0xC8;
+pub const SYSCON_IOPAD_CTRL89: u32 = SYSCON_IOPAD_CTRL_BASE + 0x164;
+pub const SYSCON_IOPAD_CTRL90: u32 = SYSCON_IOPAD_CTRL_BASE + 0x168;
+pub const SYSCON_IOPAD_CTRL91: u32 = SYSCON_IOPAD_CTRL_BASE + 0x16C;
+pub const SYSCON_IOPAD_CTRL92: u32 = SYSCON_IOPAD_CTRL_BASE + 0x170;
+pub const SYSCON_IOPAD_CTRL93: u32 = SYSCON_IOPAD_CTRL_BASE + 0x174;
+pub const SYSCON_IOPAD_CTRL94: u32 = SYSCON_IOPAD_CTRL_BASE + 0x178;
+pub const SYSCON_IOPAD_CTRL95: u32 = SYSCON_IOPAD_CTRL_BASE + 0x17C;
+pub const SYSCON_IOPAD_CTRL96: u32 = SYSCON_IOPAD_CTRL_BASE + 0x180;
+pub const SYSCON_IOPAD_CTRL97: u32 = SYSCON_IOPAD_CTRL_BASE + 0x184;
+pub const SYSCON_IOPAD_CTRL98: u32 = SYSCON_IOPAD_CTRL_BASE + 0x188;
+pub const SYSCON_IOPAD_CTRL99: u32 = SYSCON_IOPAD_CTRL_BASE + 0x18C;
+pub const SYSCON_IOPAD_CTRL100: u32 = SYSCON_IOPAD_CTRL_BASE + 0x190;
+pub const SYSCON_IOPAD_CTRL101: u32 = SYSCON_IOPAD_CTRL_BASE + 0x194;
+pub const SYSCON_IOPAD_CTRL102: u32 = SYSCON_IOPAD_CTRL_BASE + 0x198;
+pub const SYSCON_IOPAD_CTRL103: u32 = SYSCON_IOPAD_CTRL_BASE + 0x19C;
+pub const SYSCON_IOPAD_CTRL104: u32 = SYSCON_IOPAD_CTRL_BASE + 0x1A0;
+
+pub fn syscon_func_0(v: u32) {
+    // NOTE: for whatever reason, it appears that writing only works after
+    // reading i.e., if you remove the `read_32`, it breaks the code
+    // let's hope the compiler does not remove it in optimization
```

Comment:
Maybe it has something to do with cached writes or similar? Would a `fence` or other memory barrier allow for removing the `read_32`?

Not sure how much difference it makes, but might help with readability...

Reply:
Note: This is copied from the VF1 implementation and may be entirely different for the JH7110.

We are running on a single thread without interrupts here. Not sure what a memory barrier would be for. I had to read the register once, from what I found back then. May also be something else, such as alignment.

Reply:
> Not sure what a memory barrier would be for.

Right, I meant memory ordering instructions, my bad. Would `fence.i` help? It says it makes writes available to memory fetching instructions on the same hart. Not sure it would have any effect on write instructions.

Reply:
Sorry, I don't understand what this would mean here....
We just have to read that register. I do not see how that is related with anything like harts or synchronitation. It's about the behavior of the peripheral. We are on a single hart with no interrupts here.

Reply:
> Sorry, I don't understand what this would mean here....

Yeah, I was ultimately just bike-shedding here. Probably not worth it to figure out why the hardware requires the read before write. It just bugged me, because it's weird.